Bluetooth: Controller: Fix HCI command buffer allocation failure

cvinayak opened the pull request:
Fix HCI command buffer allocation failure, that can cause loss of Host Number of Completed Packets command.

~Fail by rejecting the HCI Host Buffer Size command if the required number of HCI command buffers are not allocated in the Controller implementation.~ Instead, added BUILD_ASSERT to check that allocated HCI command buffers is greater than possible generated ACL Rx buffers from the Controller when Controller to Host data flow control is supported.

Relates to commit 81614307e9fe26dbfff83848d81fd4f56eb4f14b ("Bluetooth: Add workaround
for no command buffer available")'.

Relates to commit 297f4f481f61b3f1c16ce271f4376c095797e703 ("Bluetooth: Split HCI
command & event buffers to two pools").

Relates to #81866

Modified L2CAP stress test verifies that there is no ACL Rx Buffer drops when a Heavy Work slows the Host processing and is using Controller to Host data flow control. PR with the stress test: #84602 

Another draft PR to validate the changes in this PR: #86003

cvinayak committed:
Bluetooth: Controller: Fix HCI command buffer allocation failure

Fix HCI command buffer allocation failure, that can cause
loss of Host Number of Completed Packets command.

Fail by rejecting the HCI Host Buffer Size command if the
required number of HCI command buffers are not allocated in
the Controller implementation.

When Controller to Host data flow control is supported in
the Controller only build, ensure that BT_BUF_CMD_TX_COUNT
is greater than or equal to (BT_BUF_RX_COUNT + Ncmd),
where Ncmd is supported maximum Num_HCI_Command_Packets in
the Controller implementation.

Relates to commit 81614307e9fe ("Bluetooth: Add workaround
for no command buffer available")'.

Relates to commit 297f4f481f61 ("Bluetooth: Split HCI
command & event buffers to two pools").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -184,11 +184,34 @@ config BT_BUF_CMD_TX_SIZE
+	  allocates HCI command buffers with K_NO_WAIT can end up running out
+	  of command buffers.
+
+	  When Controller to Host data flow control is enabled in the Host only
+	  build, ensure that BT_BUF_CMD_TX_COUNT is at least 2, to be able to
+	  send one normal HCI command and one Host Number of Completed Packets
+	  command.
```
Can't this be enforced via Kconfig? Something like `range 2 64 if !BT_CTLR && BT_HCI_ACL_FLOW_CONTROL`?

cvinayak replied:
Yes, `depends on !BT_HCI_RAW || !HAS_BT_CTLR` ?

I want to keep BT_BUF_CMD_TX_COUNT only for Host only and Host+Controller build, but not for Controller-only build, as it does not make sense when the Controller only supports 1 Num HCI Command Packets + BT_BUF_ACL_RX_COUNT (for Host Num Completed Packets commands).... where do you think is the right place to define BT_BUF_CMD_TX_COUNT internally, so that it is reference today in hci_raw.c and in the Controller code?

Thalley replied:
Can we keep it in the place here, but make it promptless and default to 1 for controller-only builds?

cvinayak replied:
Updated, lets see if CI is happy.

Thalley commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -184,11 +184,34 @@ config BT_BUF_CMD_TX_SIZE
+	  send one normal HCI command and one Host Number of Completed Packets
+	  command.
+
+	  When Controller to Host data flow control is supported in the
+	  Controller only build, ensure that BT_BUF_CMD_TX_COUNT is greater than
+	  or equal to (BT_BUF_ACL_RX_COUNT + Ncmd), where Ncmd is supported
+	  maximum Num_HCI_Command_Packets in the Controller implementation.
```
I don't think this can be enforced via Kconfig, but we can have BUILD_ASSERTs for it

cvinayak replied:
Yes, added runtime check on Host Buffer Size command and will derive for Controller-only (not allow it to be configured incorrectly).

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -536,10 +536,33 @@ static void host_buffer_size(struct net_buf *buf, struct net_buf **evt)
+	 * Normal HCI commands follow the HCI command flow control using
+	 * Num_HCI_Command_Packets return in HCI command complete and status.
+	 *
+	 * Note: Zephyr Controller does not support Num_HCI_Command_Packets > 1.
+	 */
+	if (acl_pkts >= CONFIG_BT_BUF_CMD_TX_COUNT) {
```
Should `CONFIG_BT_BUF_CMD_TX_COUNT` be limited to `range 1 1` for the ZLL then?

cvinayak replied:
1 when not enabling Controller to Host data flow control, and (1 + BT_BUF_ACL_RX_COUNT) when enabled.

jhedberg commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -184,11 +184,34 @@ config BT_BUF_CMD_TX_SIZE
 	help
 	  Number of buffers available for outgoing HCI commands from the Host.
 
+	  HCI Controllers may not support Num_HCI_Command_Packets > 1, hence
+	  default to 1 when not enabling Controller to Host data flow control,
+	  BT_HCI_ACL_FLOW_CONTROL.
```
This paragraph isn't completely clear to me. Even if a controller supports ncmd > 1, the Zephyr host has a command queue and will always round down the value to 1, or rather, treat it as a boolean:
https://github.com/zephyrproject-rtos/zephyr/blob/ddd795a028b1a4f9e9fa672e7791b1750f34d690/subsys/bluetooth/host/hci_core.c#L2499-L2503

The other help text paragraphs sort of make sense, but I just wanted to make sure the above host implementation detail is clear, so that this PR isn't making any false assumptions (since I haven't quite wrapped my head around all the details of it).

cvinayak replied:
Yes, Zephyr Host only implements `ncmd` = 1. See,
https://github.com/zephyrproject-rtos/zephyr/blob/ddd795a028b1a4f9e9fa672e7791b1750f34d690/subsys/bluetooth/host/hci_core.c#L4340-L4348

I guess it is not difficult to use the same semaphore and we able to support `ncmd` > 1, but we do not have support in native Controller today for that though.

cvinayak replied:
> the Zephyr host has a command queue and will always round down the value to 1, or rather, treat it as a boolean:

@jhedberg I believe the implementation "can"  support `ncmd` > 1, but today `ncmd` in the `evt` is only used as a bool to pause the queuing, but the `ncmd_sem` if initialized correctly would support `ncmd` > 1. But we are not going towards that in this PR.

jhedberg replied:
> @jhedberg I believe the implementation "can" support `ncmd` > 1, but today `ncmd` in the `evt` is only used as a bool to pause the queuing, but the `ncmd_sem` if initialized correctly would support `ncmd` > 1. But we are not going towards that in this PR.

That's not quite true, because of `bt_dev.sent_cmd`. That would need to be transformed into a list or array, which in turn would introduce some additional processing for command complete packets (right now this is O(1) when checking of the completed opcode and buffer match the last sent command (see `hci_cmd_done()` in `hci_core.c`).

cvinayak replied:
Agree. I can update the comment if required to state that even if a Controller would support `ncmd` > 1, the Host does not have the required implementation for do `ncmd` > 1.

Thalley commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,40 @@ config BT_BUF_CMD_TX_SIZE
-	range 2 64
+	int "Number of HCI command buffers" if !HAS_BT_CTLR
+	# !BT_HCI_ACL_FLOW_CONTROL ==> Permit configurable Num_HCI_Command_Packets for Host only
+	# build but hidden when HAS_BT_CTLR (default 1 for Controller only build).
```
```suggestion
	# build but hidden when HAS_BT_CTLR=n (default 1 for Controller only build).
```

cvinayak replied:
Hidden when `HAS_BT_CTLR=y`, then the `BT_BUF_CMD_TX_COUNT` be calculated.

Thalley replied:
Right, my bad :D

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -126,6 +126,12 @@ config BT_CTLR_RX_STACK_SIZE
 	int
 	default 896
 
+config BT_CTLR_HCI_CMD_TX_COUNT
+	# Hidden Controller implementation supported Num_HCI_Command_Packets value */
+	int
+	depends on BT_CTLR_HCI
+	default 1
```
Why can't we use `BT_BUF_CMD_TX_COUNT` instead?

cvinayak replied:
I should rename this to `BT_CTLR_HCI_NUM_CMD_PKT_MAX`, right? This is the supported 1 normal HCI command flow control count.

`BT_BUF_CMD_TX_COUNT` will be the buffer counts calculated with/without flow control support.

Thalley replied:
A promptless Kconfig with a static value doesn't really have any value (i.e. nonconfigurable) - It could just as well be a `#define` or something.

cvinayak replied:
As the derived value `BT_BUF_CMD_TX_COUNT` in `hci_common_internal.h` is needed by `hci_raw.c` (and a Controller implementation HCI driver glue) which will be used with other vendor Controllers (source or binary) to build Controller only builds (hci_ipc, hci_uart, etc...), I do not want to hard code using `#define` which will require `hci_raw.c` to include numerous vendor Controller's header file. Kconfig is cleaner to fetch a Controller's implementation specifics (just like we have the numerous _SUPPORT Kconfigs).

cvinayak commented:
Convert to draft, until I find out places in Host where HCI command buffers are allocated nested that are caused increased buffer requirements.

cvinayak committed:
Bluetooth: Host: Remove nested allocation of HCI command buffer

Remove nested allocation of HCI command buffer so that one
less HCI command buffer is needed starting extended
advertising with privacy enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@PavelVPV would like your review and approval on the change that could impact BT Mesh. I have tried to include mesh bsim testing on nRF5340bsim CPUAPP... but can you confirm that the failures in this PR on restricted to missing test script port towards using the `-flash_app_*`  command options to be passed to bsim, right?  (and there is no regression in other mesh tests on nrf5340bsim CPUAPP; nrf52_bsim tests have all passed).

PavelVPV commented:
> @PavelVPV would like your review and approval on the change that could impact BT Mesh. I have tried to include mesh bsim testing on nRF5340bsim CPUAPP... but can you confirm that the failures in this PR on restricted to missing test script port towards using the `-flash_app_*` command options to be passed to bsim, right? (and there is no regression in other mesh tests on nrf5340bsim CPUAPP; nrf52_bsim tests have all passed).

This is a bit difficult to say without debugging each of them. I suggest to not run mesh bsim tests on nRF53. We don't see a value in running them on nRF53 at the moment.

cvinayak commented:
> > @PavelVPV would like your review and approval on the change that could impact BT Mesh. I have tried to include mesh bsim testing on nRF5340bsim CPUAPP... but can you confirm that the failures in this PR on restricted to missing test script port towards using the `-flash_app_*` command options to be passed to bsim, right? (and there is no regression in other mesh tests on nrf5340bsim CPUAPP; nrf52_bsim tests have all passed).
> 
> This is a bit difficult to say without debugging each of them. I suggest to not run mesh bsim tests on nRF53. We don't see a value in running them on nRF53 at the moment.

As there are no Host bsim tests on nrf5340 APP core today (and a lot of work to update all the testcases.yml file to add it), mesh tests where a guinea pig to check if I would have introduced any regressions with my changed to how HCI command buffers are allocated. The nrf52_bsim tests passed, and the nrf5340bsim CPU app failures mostly look it due to incorrect bsim flash option used.

Yes, I will remove the commit to enable mesh testing on nRF5340bsim app core, and it can be a separate task later.

cvinayak commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,45 @@ config BT_BUF_CMD_TX_SIZE
+	# !BT_HCI_ACL_FLOW_CONTROL ==> Permit configurable Num_HCI_Command_Packets for
+	# Host only build but hidden when HAS_BT_CTLR=y (default 1 for Controller only build).
+	# !(BT_HCI_RAW && HAS_BT_CTLR) ==> Host only build.
+	# BT_MESH is starting Extended Advertising in callback run in system work queue, hence
```
Done.

cvinayak commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,45 @@ config BT_BUF_CMD_TX_SIZE
 	  Number of buffers available for outgoing HCI commands from the Host.
 
+	  HCI Controllers may not support Num_HCI_Command_Packets > 1, hence
+	  defaults to 1 when not enabling Controller to Host data flow control
```
done

cvinayak commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,45 @@ config BT_BUF_CMD_TX_SIZE
+	  Normal HCI commands follow the HCI command flow control using Num_HCI_Command_Packets
+	  return in HCI command complete and status.
+
+	  Host Number of Completed Packets command does not follow normal flow control of HCI
```
done

cvinayak commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,45 @@ config BT_BUF_CMD_TX_SIZE
+	  commands and the Controller side HCI drivers that allocates HCI command buffers with
+	  K_NO_WAIT can end up running out of command buffers.
+
+	  When Controller to Host data flow control is enabled in the Host only or combined
```
Done.

cvinayak commented on subsys/bluetooth/common/hci_common_internal.h:
```diff
@@ -0,0 +1,83 @@
+#if !defined(CONFIG_BT_HCI_RAW) || !defined(CONFIG_HAS_BT_CTLR) || \
+	!defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
+#if !defined(CONFIG_BT_HCI_RAW)
+/* Auto initiated additional HCI command buffers enqueued in the Host */
```
done

cvinayak commented on subsys/bluetooth/common/hci_common_internal.h:
```diff
@@ -0,0 +1,83 @@
+#endif /* !CONFIG_BT_HCI_RAW */
+
+/* When Controller to Host data flow control is supported in Host plus Controller build,
+ * Host only build, then we need additional BT_BUF_ACL_RX_COUNT number of HCI command buffers for
```
done

jhedberg commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,44 @@ config BT_BUF_CMD_TX_SIZE
+	# !BT_HCI_ACL_FLOW_CONTROL ==> Permit configurable Num_HCI_Command_Packets for
+	# Host-only build but hidden when HAS_BT_CTLR=y (default 1 for Controller-only build).
+	# !(BT_HCI_RAW && HAS_BT_CTLR) ==> Host-only build.
+	depends on !BT_HCI_RAW || !HAS_BT_CTLR || !BT_HCI_ACL_FLOW_CONTROL
```
At least in my mind, it's easier to parse `!(A && B && C)` than `(!A || !B || !C)`. Maybe I'm alone in that, though :)

Thalley replied:
I do actually prefer the latter :D That way you only, in your mind, have to parse a single value at a time :D Might be related to brain cache size :P

jhedberg replied:
Ok. Fair enough :)

jhedberg commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +183,44 @@ config BT_BUF_CMD_TX_SIZE
-	int "Number of HCI command buffers"
-	default 2
-	range 2 64
+	int "Number of HCI command buffers" if !BT_HCI_RAW || !HAS_BT_CTLR
```
Same here.

jhedberg commented on subsys/bluetooth/common/hci_common_internal.h:
```diff
@@ -0,0 +1,84 @@
+#define BT_BUF_CMD_TX_HOST_NUM_CMPLT_PKT 0
+#endif /* !CONFIG_BT_HCI_ACL_FLOW_CONTROL */
+
+/* Based on Host + Controller, Host-only or Controller-only */
+#define BT_BUF_CMD_TX_COUNT (CONFIG_BT_BUF_CMD_TX_COUNT + \
+			     BT_BUF_CMD_TX_HOST_NUM_CMPLT_PKT + \
+			     BT_BUF_CMD_TX_REMOTE_VERSION + \
+			     BT_BUF_CMD_TX_AUTO_PHY_UPDATE + \
+			     BT_BUF_CMD_TX_AUTO_DATA_LEN_UPDATE)
```
The one thing that has me a little bit concerned with all this logic, is that I'd like our implementation to be flexible enough to easily handle a future redesign where the memory management of TX buffers would completely or partially be outside of the Bluetooth stack, e.g. the app could just have its own buffer pools and use buffers from there to send HCI commands.

Or to put it another way, the handling of buffers should be such that, as long as some minimum is guaranteed, it doesn't matter how many buffers are alive (allocated) at any given moment. If the controller can't handle more than a certain amount, the rest will just sit in a command queue on the host side.

cvinayak replied:
> the handling of buffers should be such that, as long as some minimum is guaranteed, it doesn't matter how many buffers are alive (allocated) at any given moment.

It is not possible now, when we are using system work queue (`CONFIG_BT_RECV_WORKQ_SYS`), to have a minimum guaranteed as it is not allowed to block for buffer availability.

Even with the minimum calculated as above based on auto-initiated procedures that happens in the system work queue, there can be any number of additional HCI command buffers required at runtime due to our choice to have callback invoked from system work queue. BT Mesh has been facing this challenge, and we are conveniently advising applications to increase `CONFIG_BT_BUF_CMD_TX_COUNT` :-(

> If the controller can't handle more than a certain amount, the rest will just sit in a command queue on the host side.

This PR is not about the Controller's support for handling Num_HCI_Command_Packets > 1, but about the lost ability in Host to manage a command queue inherently in the form of waiting threads (maintain the suggested workaround of increasing `CONFIG_BT_BUF_CMD_TX_COUNT`) and try gracefully supporting  Host Number of Completed Packets command. 

I know where you and I come from, but the Host is not what it is that I remember!

jhedberg replied:
@cvinayak thanks for the clarification. It's all starting to make a bit more sense to me, and also showing me that I should really get more familiar with the new design philosophy of the host.

cvinayak commented:
@jhedberg  and @alwa-nordic this PR was a slipper slope, but HCI command and data flow control should be resolved now for the right buffer allocations. The event Rx flow control will have dependency still on the transport level flow control.

Tranport level flow control will be manadatory for the case for Host and Controller split applications. For Host plus Controller application the inherent function interface based control flow will ensure the event rx does not overflow (discardable event being discarded too as a mitigation). Any deadlocks due to overwhelming event rx should be address by adding new testcases to cover it and be fixed as necessary in a separate PR.

alwa-nordic reviewed (dismissed):
> Tranport level flow control will be manadatory for the case for Host and Controller split applications.

I disagree with this. I want split and non-split builds to work the same way. And I insist the HCI transport has nothing to do with the contents of the HCI messages.

To me, ACL flow control is not for the benefit of the HCI transport. It's optional after all. It's for the application in case it needs to handle events (like command complete) during handling of ACL data.

@PavelVPV and I believe we have an alternative solution, so we would like to discuss this with you further. I'll NAK this for now so we don't merge it by accident.

Thalley commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -117,10 +117,12 @@ BUILD_ASSERT(BT_BUF_ACL_RX_COUNT <= BT_BUF_ACL_RX_COUNT_MAX,
-#define BT_BUF_RX_COUNT (MAX(MAX(CONFIG_BT_BUF_EVT_RX_COUNT, \
-				 BT_BUF_ACL_RX_COUNT), \
-			     BT_BUF_ISO_RX_COUNT))
+BUILD_ASSERT(CONFIG_BT_BUF_EVT_RX_COUNT > CONFIG_BT_BUF_ACL_TX_COUNT,
+	     "Increase Event RX buffer count to be greater than ACL TX buffer count");
```
Can you expand with a comment on why?

Thalley commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -79,14 +79,20 @@ config BT_BUF_ACL_RX_SIZE
 
 config BT_BUF_ACL_RX_COUNT_EXTRA
 	int "Number of extra incoming ACL data buffers"
-	default 0
+	range 1 65535 if BT_CONN
 	range 0 65535
+	default BT_MAX_CONN if BT_CONN
+	default 0
```
Not sure I follow this: 
```
default BT_MAX_CONN if BT_CONN
```

If `BT_CONN=n`, then why we do have any ACL buffers?

Thalley replied:
Ah, yeah that makes sense. Could you add that as a comment for the Kconfig option?

Thalley commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -79,14 +79,20 @@ config BT_BUF_ACL_RX_SIZE
 
-	  By default, the number of incoming ACL data buffers is equal to
-	  CONFIG_BT_MAX_CONN + 1.
+	  By default, the total number of incoming ACL data buffers is equal to
+	  CONFIG_BT_MAX_CONN + 1, to support L2CAP recombination.
```
Isn't this incorrect with `default BT_MAX_CONN if BT_CONN` above?

Thalley commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +189,44 @@ config BT_BUF_CMD_TX_SIZE
+	# Host-only build but hidden when HAS_BT_CTLR=y (default 1 for Controller-only build).
+	# !(BT_HCI_RAW && HAS_BT_CTLR) ==> Host-only build.
+	depends on !BT_HCI_RAW || !HAS_BT_CTLR || !BT_HCI_ACL_FLOW_CONTROL
+	# BT_MESH is starting Extended Advertising in the callback run in the system work queue and
+	# as the Host also uses the system work queue for Tx towards the Controller, one additional
+	# HCI command buffer is needed.
+	range 2 64 if BT_MESH
```
Can you elaborate on this? 

`system work queue for Tx towards the Controller` - TX of what? 
Aren't the HCI commands sent synchronously?

cvinayak commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -97,13 +97,13 @@ struct bt_buf_data {
-	(MAX(CONFIG_BT_BUF_ACL_RX_COUNT, (CONFIG_BT_MAX_CONN + 1)) +                               \
-	 CONFIG_BT_BUF_ACL_RX_COUNT_EXTRA)
+#define BT_BUF_ACL_RX_COUNT_EXTRA CONFIG_BT_BUF_ACL_RX_COUNT_EXTRA
+#define BT_BUF_ACL_RX_COUNT       (MAX(CONFIG_BT_BUF_ACL_RX_COUNT, 1) + BT_BUF_ACL_RX_COUNT_EXTRA)
```
Still thinking on this, `BT_BUF_HCI_ACL_RX_COUNT` introduced in this PR is ending up being `1` always! 

Thinking aloud,
`BT_BUF_HCI_ACL_RX_COUNT` should able to be greater than or equal to 1, so that more number of ACL buffers be buffered up for host processing (in bursts, x bps) to match the reception over transport (in bursts, y bps).

At the same time, `BT_BUF_ACL_RX_COUNT_EXTRA` be for L2CAP recombination, so that `BT_BUF_HCI_ACL_RX_COUNT`  is available for ACL buffer with and without Controller to Host data flow control.

cvinayak replied:
Created for follow up: https://github.com/zephyrproject-rtos/zephyr/issues/86142

Thalley commented on subsys/bluetooth/common/hci_common_internal.h:
```diff
@@ -0,0 +1,108 @@
+
+/* Auto initiated additional HCI command buffers enqueued in the Host */
+
+#if defined(CONFIG_BT_REMOTE_VERSION)
+#define BT_BUF_CMD_TX_REMOTE_VERSION       1
+#else /* !CONFIG_BT_REMOTE_VERSION */
+#define BT_BUF_CMD_TX_REMOTE_VERSION       0
+#endif /* !CONFIG_BT_REMOTE_VERSION */
+
+#if defined(CONFIG_BT_AUTO_PHY_UPDATE)
+#define BT_BUF_CMD_TX_AUTO_PHY_UPDATE      1
+#else /* !CONFIG_BT_AUTO_PHY_UPDATE */
+#define BT_BUF_CMD_TX_AUTO_PHY_UPDATE      0
+#endif /* !CONFIG_BT_AUTO_PHY_UPDATE */
+
+#if defined(CONFIG_BT_AUTO_DATA_LEN_UPDATE)
+#define BT_BUF_CMD_TX_AUTO_DATA_LEN_UPDATE 1
+#else /* !CONFIG_BT_AUTO_DATA_LEN_UPDATE */
+#define BT_BUF_CMD_TX_AUTO_DATA_LEN_UPDATE 0
+#endif /* !CONFIG_BT_AUTO_DATA_LEN_UPDATE */
```
I think some of these can be better replaced with `COND_CODE_1` to make it simpler:

e.g. 
```
#if defined(CONFIG_BT_REMOTE_VERSION)
#define BT_BUF_CMD_TX_REMOTE_VERSION       1
#else /* !CONFIG_BT_REMOTE_VERSION */
#define BT_BUF_CMD_TX_REMOTE_VERSION       0
#endif /* !CONFIG_BT_REMOTE_VERSION */
```

becomes 
```
#define BT_BUF_CMD_TX_REMOTE_VERSION COND_CODE_1(CONFIG_BT_REMOTE_VERSION, (1), (0))`
```

cvinayak replied:
Done

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -21,6 +21,8 @@
 #include <zephyr/bluetooth/hci_vs.h>
 #include <zephyr/bluetooth/buf.h>
 
+#include "../common/hci_common_internal.h"
```
```suggestion
#include "common/hci_common_internal.h"
```

`common` should be in the include path

Thalley commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -97,13 +97,13 @@ struct bt_buf_data {
-	(MAX(CONFIG_BT_BUF_ACL_RX_COUNT, (CONFIG_BT_MAX_CONN + 1)) +                               \
-	 CONFIG_BT_BUF_ACL_RX_COUNT_EXTRA)
+#define BT_BUF_ACL_RX_COUNT_EXTRA CONFIG_BT_BUF_ACL_RX_COUNT_EXTRA
+#define BT_BUF_ACL_RX_COUNT       (MAX(CONFIG_BT_BUF_ACL_RX_COUNT, 1) + BT_BUF_ACL_RX_COUNT_EXTRA)
 #else
-#define BT_BUF_ACL_RX_COUNT 0
+#define BT_BUF_ACL_RX_COUNT_EXTRA 0
+#define BT_BUF_ACL_RX_COUNT       0
 #endif /* CONFIG_BT_CONN && CONFIG_BT_HCI_HOST */
```
Consider using `COND_CODE_1` instead of `#if ... #else ... #endif`

Thalley reviewed (commented):
@cvinayak any further changes pending, or ready for re-review?

cvinayak commented:
> @cvinayak any further changes pending, or ready for re-review?

Almost, just need to validate this PR over a test that may need it.

@PavelVPV do you have the PR submitted that will depend on this PR (I can have draft PR to check if pass in the CI) ?

PavelVPV commented:
> > @cvinayak any further changes pending, or ready for re-review?
> 
> Almost, just need to validate this PR over a test that may need it.
> 
> @PavelVPV do you have the PR submitted that will depend on this PR (I can have draft PR to check if pass in the CI) ?

I'll post a PR later (I hope today).

PavelVPV commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +191,44 @@ config BT_BUF_CMD_TX_SIZE
+	# Host-only build but hidden when HAS_BT_CTLR=y (default 1 for Controller-only build).
+	# !(BT_HCI_RAW && HAS_BT_CTLR) ==> Host-only build.
+	depends on !BT_HCI_RAW || !HAS_BT_CTLR || !BT_HCI_ACL_FLOW_CONTROL
+	# BT_MESH is starting Extended Advertising in the callback run in the system work queue and
+	# as the Host also uses the system work queue for Tx towards the Controller, one additional
+	# HCI command buffer is needed.
+	range 2 64 if BT_MESH
```
This is not fully correct any more because mesh now by default uses a separate workq:
https://github.com/zephyrproject-rtos/zephyr/blob/e885c8c42f775f92176e7ad7eb71d8dcdc2bb4e0/subsys/bluetooth/mesh/Kconfig#L106-L135

I suggest to set limit range for `BT_MESH_WORKQ_SYS`:
```suggestion
	# BT_MESH when compiled with BT_MESH_WORKQ_SYS is starting Extended Advertising in the callback
	# run in the system work queue and as the Host also uses the system work queue for Tx towards
	# the Controller, one additional HCI command buffer is needed.
	range 2 64 if BT_MESH_WORKQ_SYS
```

PavelVPV commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +191,44 @@ config BT_BUF_CMD_TX_SIZE
-	int "Number of HCI command buffers"
-	default 2
-	range 2 64
+	int "Number of HCI command buffers" if !BT_HCI_RAW || !HAS_BT_CTLR
+	# !BT_HCI_ACL_FLOW_CONTROL ==> Permit configurable Num_HCI_Command_Packets for
+	# Host-only build but hidden when HAS_BT_CTLR=y (default 1 for Controller-only build).
+	# !(BT_HCI_RAW && HAS_BT_CTLR) ==> Host-only build.
+	depends on !BT_HCI_RAW || !HAS_BT_CTLR || !BT_HCI_ACL_FLOW_CONTROL
```
This is a bit difficult to read. Maybe something like this?
```suggestion
	# This option is only visible for a user when Host and Controller are build together, or for
	# Host-only builds.
	int "Number of HCI command buffers" if !BT_HCI_RAW || !HAS_BT_CTLR
	# This option is present when Host and Controller are build together, or
	# for Host only builds, or when Controller-to-Host ACL data flow control
	# is disabled.
	depends on !BT_HCI_RAW || !HAS_BT_CTLR || !BT_HCI_ACL_FLOW_CONTROL
```

cvinayak commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -183,12 +191,46 @@ config BT_BUF_CMD_TX_SIZE
+	# for Host only builds, or when Controller-to-Host ACL data flow control
+	# is disabled.
+	depends on !BT_HCI_RAW || !HAS_BT_CTLR || !BT_HCI_ACL_FLOW_CONTROL
+	# BT_MESH when compiled with BT_MESH_WORKQ_SYS is starting Extended Advertising in the callback
+	# run in the system work queue and as the Host also uses the system work queue for Tx towards
+	# the Controller, one additional HCI command buffer is needed.
+	range 2 64 if BT_MESH_WORKQ_SYS
+	range 1 64
+	default 2 if BT_MESH
+	default 1
```
@PavelVPV  I took your suggestion; Looks like `BT_MESH` always requires atleast `2`, some of the tests failed when using `1` when `BT_MESH` and `!BT_MESH_WORKQ_SYS` ... is this correct?

PavelVPV replied:
Ah, I see. I checked this. While some test explicitly require 2 cmd buffers because they do 2 things in parallel (start advertising and scanning), `pb_remote_reprovision` failed because this problem with tx thread removal in the host hit mesh again. This time it is because the host's tx processor that runs on sysworkq is blocked by [the mesh node reset](https://github.com/zephyrproject-rtos/zephyr/blob/a9b76dcb1ba509f58041ca49542839021b8e6ec7/subsys/bluetooth/mesh/cfg_srv.c#L43-L46) functionality. I could fix this specific issue, but we have suspend/resume API that can also be called from sysworkq thus causing this deadlock anyway. I'll follow up on this issue separately. So for now, keep `if BT_MESH` for `default` and `range`.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -125,11 +125,14 @@ config BT_MESH_WORKQ_SYS
-	  CONFIG_BT_BUF_CMD_TX_COUNT, the host returns -ENOBUFS immediately
-	  and the mesh drops the message transmission. To mitigate this
-	  issue, make sure to have sufficient number of HCI command buffers.
+	  When this option is enabled and the mesh tries to send a message, the
+	  Host can run out of the HCI command buffers controlled by the
+	  CONFIG_BT_BUF_CMD_TX_COUNT value and the Host returns -ENOBUFS
+	  immediately and the mesh drops the message transmission.
+
+	  To mitigate dropping of messages, make sure to have sufficient number
+	  of HCI command buffers (CONFIG_BT_BUF_CMD_TX_COUNT).
```
We are updating this note here: https://github.com/zephyrproject-rtos/zephyr/pull/86022/files. I think you can remove this change.

cvinayak commented:
> Discussion has been had. I still believe what I wrote in the NAK, but fixing it in a different way will expose deadlocks elsewhere. Though I don't believe this is a step in the right direction design-wise, I've been convinced to be pragmatic and allow merging this to resolve some immediate issues.

This PR is not solving the scenario wherein HCI Event will be dropped; only HCI Rx buffer count are fixed to mitigate deadlocks related to processing of received ACL Data versus Events from the Controller when Controller to Host Flow Control is in use. Accordingly, FIXME code comment has been added.
